File: Dev/test_cs/ObjectSystem/Component.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ace;

namespace test_cs.ObjectSystem
{
	class MyComponent : ace.Object2DComponent
	{
		public MyComponent()
		{
			time = 0;
		}

		protected override void OnUpdate()
		{
			if( time % 30 == 0 )
			{
				Owner.Angle += 13;
			}
			++time;
		}

		private int time { get; set; }
	}

	class Component : ISample
	{
		public void Run()
		{
			Engine.Initialize( "Ace Component", 640, 480, new EngineOption() );

			var scene = new Scene();
			var layer = new Layer2D();
			var obj = new TextureObject2D()
			{
				Position = new Vector2DF( 320, 240 ),
				Texture = Engine.Graphics.CreateTexture2D( "Data/Texture/Cloud1.png" ),
			};

			layer.AddObject( obj );
			scene.AddLayer( layer );
			Engine.ChangeScene( scene );
			obj.AddComponent( new MyComponent(), "rotation" );

			while( Engine.DoEvents() )
			{
				Engine.Update();
			}

			Engine.Terminate();
		}
	}
}

File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/Component.cpp
#include <Altseed.h>
#include <gtest/gtest.h>
#include <memory>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class MyObject2DComponent : public asd::Object2DComponent
{
private:
	int time;

public:
	MyObject2DComponent()
		: time(0)
	{
	}

protected:
	void OnUpdate()
	{
		if (time % 6 == 0)
		{
			GetOwner()->SetAngle(GetOwner()->GetAngle() + 13);
		}
		++time;
	}
};

class ObjectSystem_Component : public EngineTest
{
public:
	ObjectSystem_Component(bool isOpenGLMode)
		: EngineTest(asd::ToAString("Component"), isOpenGLMode, 30)
	{
	}

protected:
	void OnStart()
	{
		auto scene = make_shared<Scene>();
		auto layer = make_shared<Layer2D>();
		auto object = make_shared<TextureObject2D>();
		auto component = make_shared<MyObject2DComponent>();

		asd::Engine::ChangeScene(scene);
		scene->AddLayer(layer);
		layer->AddObject(object);

		object->SetPosition(Vector2DF(320, 240));
		object->AddComponent(component, ToAString("rotation"));
		//ASSERT_NE(object->GetComponent(ToAString("rotation")), nullptr);

		auto g = asd::Engine::GetGraphics()->CreateTexture2D(ToAString("Data/Texture/Cloud1.png").c_str());
		object->SetTexture(g);
	}
};

ENGINE_TEST(ObjectSystem, Component)
